libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -599,6 +538,135 @@ void AC_AttitudeControl::input_angle_step_bf_roll_pitch_yaw(float roll_angle_ste
     attitude_controller_run_quat();
 }
 
+// Command a Quaternion attitude with feedforward and smoothing
+void AC_AttitudeControl::input_thrust_vector_rate_heading(Vector3f thrust_vector, float heading_rate_cds)
+{
+    // Convert from centidegrees on public interface to radians
+    const float heading_rate = radians(heading_rate_cds * 0.01f);
+
+    // calculate the attitude target euler angles
+    _attitude_target.to_euler(_euler_angle_target.x, _euler_angle_target.y, _euler_angle_target.z);
+
+    thrust_vector.normalize();
+
+    // the cross product of the desired and target thrust vector defines the rotation vector
+    Vector3f thrust_vec_cross = Vector3f{0.0f,0.0f,-1.0f} % thrust_vector;
+
+    // the dot product is used to calculate the angle between the target and desired thrust vectors
+    const float thrust_vector_angle = acosf(constrain_float(Vector3f(0.0f,0.0f,-1.0f) * thrust_vector, -1.0f, 1.0f));
```

Comment:
A non-functional change but maybe we should create a local variable to avoid the duplicate, "Vector3f(0.0f,0.0f,-1.0f)"?  It's not a big deal of course because creating it is extremely quick.

const Vector3f thrust_vector_up{0.0f, 0.0f, -1.0f};

Reply:
Nice, done.